src/NumberTheory/vinberg.jl

```diff
@@ -353,3 +353,163 @@ function vinberg_algorithm(S::ZZLat, upper_bound; v0=QQ[0;]::QQMatrix, root_leng
+end
+
+function _get_C0(Lf::ZZLatWithIsom, tau::QQBarFieldElem)::PolyRingElem
+  charPolyF = characteristic_polynomial(Lf)
+  x = gen(parent(charPolyF))
+  (n, remainder) = remove(charPolyF, x-1) # remove all x-1 factors
+  return div(remainder, minpoly(parent(charPolyF), tau)) #remove Salem polynomial of salem number tau
+end
+
+function _get_Cfancy(Lf::ZZLatWithIsom, C0)
+  return short_vectors_iterator(lattice(kernel_lattice(Lf, C0)), 2 , 2)
+end
+
+
+@doc raw"""
+  _get_h(L::ZZLat, v, w, Qb, bi_form)->QQMatrix
+
+THe function returns $h$ vector, such that $h^2>0$ in ambient space of 'L' using 'bi_form' as a product function. 
+
+Vectors 'v','w' are eigenvectors of isometry assotiated with L based on salem number $tau$ and $tau^(-1)$
+
+'Qb' is a current field, where all calculations happens
+"""
+function _get_h(L::ZZLat, v, w, Qb, bi_form)::QQMatrix
+  RR = ArbField(64)
+  l = number_of_rows(basis_matrix(L))
+  h0 = (v+w)*change_base_ring(Qb, inv(basis_matrix(L)))
+  n = 1
+  i = 1
+  z = matrix(Qb,1,l,rand(-10:10, l)) # random small vector in lattice basis
+  h = (z+h0*n) # in lattice basis
+  h = map(x->QQ(ZZ(floor(RR(x)))) , h)*basis_matrix(L) # in ambient space basis and rounded
+  while bi_form(h,h) <= 0 || n == 10000  # block on 10000, because for a bigger n time consumption is way too big
+    i+=1
+    if i == 100
+      n+=1
+      i = 1
+    end
+    h = (z+h0*n) #in lattice basis
+    h = map(x->QQ(ZZ(floor(RR(x)))) , h)*basis_matrix(L) # in ambient space basis and rounded
+  end
+  return h
+end
+
+function _get_R(L::ZZLat, h::QQMatrix)::Vector{QQMatrix}
+  return short_vectors_affine(L,h,0,-2) #better to use iterator, but there is no equivalent function that returns iterator
+end
+
+@doc raw"""
+  _process_finite_sets_of_h(h, f::QQMatrix, v, w, bi_form, L::ZZLat) -> Tuple{Bool, QQMatrix}
+
+The function calculates all $r$, that can be obstructing roots of 'L'.
+Calculation is made one by one and then the $r$ is checked
+$r$ is based on pairs of integer $(a,b)$ of $-2x^2+2y^2+2aby>=x(a^2+b^2)$ with $a>0$, $b<0$, $x>0$,$y>0$
+See Steps 7,8,9 of Algorithm 5.8 of "MINIMUM POSITIVE ENTROPY OF COMPLEX
+ENRIQUES SURFACE AUTOMORPHISMS" by KEIJI OGUISO and XUN YU in Duke Mathematical Journal, 2020,
+volume 169, number 18, pages 3565 - 3606
+
+Return a tuple of a boolean that represents if isometry f of the lattice is positive and QQMatrix,
+that represents an obstructing root
+"""
+function _process_finite_sets_of_h(h, f::QQMatrix, v, w, bi_form, L::ZZLat):: Tuple{Bool, QQMatrix}
+  x = bi_form(h, h)
+  y = bi_form(h, h*f)
+  z = y^2-x^2
+  if (z<0) return (true, zero(h)) end #then discrininant for a will be <0 and there is no roots (a,b)=> no obstructing roots => positive
+  #trunc(ZZRingElem,-sqrt(2*z/x)) #check issues on github as round works in the same context
+  b_min = round(ZZRingElem, -sqrt(2*z/x), RoundUp)
+  for b = b_min:-1
+    a_max = round(ZZRingElem, (ZZ(b)*y+sqrt(z*(ZZ(b)^2+2*x)))/x, RoundDown)
```

Comment:
```suggestion
    a_max = b*y+isqrt(z*(b^2+2*x)/x^2)
```
Note the `ZZ(x)*y` is not necessary since `x` will be coerced (promoted) to `QQFieldElem` automatically (if `y isa QQFieldElem`)

Reply:
Made a slightly different commit to round down inside `isqrt`